Automattic/wp-calypso
# Replace chokidar-cli package watcher with node version

#### Changes proposed in this Pull Request

This is a follow-up to https://github.com/Automattic/wp-calypso/pull/42867 which made it so that whenever a change occurs to a file in the `packages` directory, only that package is rebuilt. This PR makes two improvements to that one:

1. If there are multiple files changed at once (eg: when changing branches), the old script only rebuilds the most recent package. This script will rebuild all changed packages (but only once per package).
2. The old script (and the script before that) uses chokidar-cli and file globs which (I think) limits it to UNIX-like OSes. This script uses the chokidar npm library directly and performs watching on its own using the node `path` library to safely deal with path separators on different OSes.

Props to @alshakero for the idea.

#### Testing instructions

- Run `yarn start` and wait for calypso to be started.
- Modify a source file inside the `packages` directory.
- Verify that you see a `rebuilding` message for the package which contains the changed file. 
- Verify that the change you made took effect by loading http://calypso.localhost:3000 and looking for your change. A good change to make is a `console.log()` statement for which it is easy to see the results.
- Pause webpack and the watcher (<kbd>CTRL-z</kbd> on UNIX-like OSes).
- Make changes to several files as above, but make the changes in at least two different packages and edit two files within a single package. For example: `touch packages/composite-checkout/src/components/field.js packages/load-script/src/index.js packages/composite-checkout/src/public-api.js`
- Resume the build process (`fg` on UNIX-like OSes).
- Verify that you see a `rebuilding` message once for each package which contains a changed file.
